### PR TITLE
Adding CIBW_BUILD option

### PR DESCRIPTION
**Pull request by YannickJadoul**
Cfr. #98, this is adding CIBW_BUILD to positively select a set of build targets (rather than deselecting everything else).

I'm not completely happy with the naming of `BuildSelector` and `selection` (replacing `BuildSkipper` and the `skip` argument to the different platforms), but I can't think of anything better, right now.

Comments, remarks, and identification of bugs are very welcome, @altendky, @astrofrog, and @joerick!

### CODE REVIEW

**Comment by altendky** [comment]
Thanks. Quick look over the code mostly makes sense. I'll give it a shot over in twisted/twisted#1067.

File: README.md [hunk 1]
```diff
@@ -213,15 +216,16 @@ Platform tags look like `macosx_10_6_intel` `manylinux1_x86_64` `manylinux1_i686
...
-- Skip building on Python 2.7 on the Mac: `cp27-macosx_10_6_intel`  
-- Skip building on Python 2.7 on all platforms: `cp27-*`  
-- Skip Python 2.7 on Windows: `cp27-win*`  
-- Skip Python 2.7 on 32bit Windows: `cp27-win32`  
-- Skip Python 3.4 and Python 3.5: `cp34-* cp35-*`
-- Skip Python 3.6 on Linux: `cp36-manylinux*`
-- Only build on Python 3.6: `cp27-* cp34-* cp35-*`
-
-***
+- Only build on Python 3.6: `CIBW_BUILD`:`cp36-*`
```
Comment (joerick) [inline README.md]: ohhhh CIBW_BUILD can do wildcards too! 🤯that's nice

File: README.md [hunk 2]
```diff
@@ -213,15 +216,16 @@ Platform tags look like `macosx_10_6_intel` `manylinux1_x86_64` `manylinux1_i686
...
+- Skip building on Python 2.7 on the Mac: `CIBW_SKIP`:`cp27-macosx_10_6_intel`
+- Skip building on Python 2.7 on all platforms: `CIBW_SKIP`:`cp27-*`
+- Skip Python 2.7 on Windows: `CIBW_SKIP`:`cp27-win*`
+- Skip Python 2.7 on 32-bit Windows: `CIBW_SKIP`:`cp27-win32`
+- Skip Python 3.4 and Python 3.5: `CIBW_SKIP`:`cp34-* cp35-*`
+- Skip Python 3.6 on Linux: `CIBW_SKIP`:`cp36-manylinux*`
+- Only build on Python 3 and skip 32-bit builds: `CIBW_BUILD`:`cp3?-*` and `CIBW_SKIP`:`*-win32 *-manylinux1_i686`
```
Comment (joerick) [inline README.md]: nice :)

File: cibuildwheel/__main__.py [hunk 3]
```diff
@@ -99,7 +99,7 @@ def main():
         traceback.print_exc(None, sys.stderr)
         exit(2)
 
-    skip = BuildSkipper(skip_config)
+    selection = BuildSelector(build_config, skip_config)
```
Comment (joerick) [inline cibuildwheel/__main__.py]: BuildSelector seems like a reasonable name to me. `selection` is a little confusing though? Maybe just call it `build_selector`? open to other ideas though

File: cibuildwheel/util.py [hunk 4]
```diff
@@ -21,15 +21,18 @@ def get_build_verbosity_extra_flags(level):
...
+class BuildSelector(object):
+    def __init__(self, build_config, skip_config):
+        self.build_patterns = build_config.split()
+        self.skip_patterns = skip_config.split()
 
     def __call__(self, build_id):
-        return any(fnmatch(build_id, pattern) for pattern in self.patterns)
+        def match_any(patterns):
+            return any(fnmatch(build_id, pattern) for pattern in patterns)
+        return match_any(self.build_patterns) and not match_any(self.skip_patterns)
```
Comment (joerick) [inline cibuildwheel/util.py]: So the logic is 'match any of the build patterns, and none of the skip patterns.' Makes sense!

> (b1 or b2 or b3) and not (s1 or s2 or s3)  
> (b1 or b2 or b3) and not s1 and not s2 and not s3

File: test/04_build_skip/environment.json [hunk 5]
```diff
@@ -0,0 +1,4 @@
+{
+    "CIBW_BUILD": "cp3?-*",
+    "CIBW_SKIP": "cp33-*"
```
Comment (joerick) [inline test/04_build_skip/environment.json]: Python 3.3 is about to be removed from cibuildwheel - maybe better to change this to Python 3.4

File: README.md (hunk 1 above)
Comment (YannickJadoul) [reply README.md]: Your original code, just duplicated ;-)

File: cibuildwheel/util.py (hunk 4 above)
Comment (YannickJadoul) [reply cibuildwheel/util.py]: Credit to @altendky for coming up with those semantics. But indeed, I've been thinking a little bit more, and this gives a really intuitive interpretation (there's not even really a 'first `CIBW_BUILD` is applied, and then `CIBW_SKIP`' kind of sequential semantics)

File: cibuildwheel/__main__.py (hunk 3 above)
Comment (YannickJadoul) [reply cibuildwheel/__main__.py]: Myeah, I liked how `skip(...)` read, but couldn't really reproduce that

Comment (YannickJadoul) [reply cibuildwheel/__main__.py]: The other option is to keep this as `BuildSkipper`, btw, and just add the extra argument of `build_config`? Depending on your point of view, this class is still determining what builds to skip.

Comment (joerick) [reply cibuildwheel/__main__.py]: BuildSkipper feels like it relates too closely with CIBW_SKIP. I like it the way it is, I think, just the `selection(identifier)` is a bit too vague.

**Comment by altendky** [comment]
https://travis-ci.org/twisted/twisted/builds/429309631?utm_source=github_status&utm_medium=notification

It's a pretty minimal 'test' but I think it did what I wanted.

File: test/04_build_skip/environment.json (hunk 5 above)
Comment (YannickJadoul) [reply test/04_build_skip/environment.json]: Changed the folder name anyway, and added the line. So to almost any diff, this file is (almost) completely changed anyway, so might as well clean this up? :-/

**Comment by YannickJadoul** [comment]
Tests are failing because on OSX, there's now only 3 wheels being produced for the `04_build_skip` test (i.e., 3.5, 3.6, 3.7, because `CIBW_BUILD` selects only 3.<something> and `CIBW_SKIP` skips 3.4, and because 3.3 is already not supported anymore on macOS). I'm seeing if I can find a solution.

**Comment by YannickJadoul** [comment]
> I'm seeing if I can find a solution.

Pfff, I wanted to add a file with the list of files to be produced by a test, but then I need to start detecting the platform, as well.

I could split the test again into one for `SKIP` and one for `BUILD`, I suppose, but ... meh :-/ Any suggestions on making this test and its assert work are welcome.

**Comment by joerick** [comment]
> Any suggestions on making this test and its assert work are welcome.

Hmm. We've talked in the past of the need to update how these tests work with a .py file for each script. Maybe we create an issue to track that refactoring, and for now we can reduce that magic number to 3 (there's nothing special about minimum 4 wheels per platform).

**Comment by altendky** [comment]
Do we need more review here?  Trying it out in twisted/twisted#1067.

**Comment by altendky** [comment]
Yeah, I'm still happy with it from what I've checked.  @joerick, do you see anything else we need to address?

**Comment by joerick** [comment]
Released in 0.10.0! Thanks both!

**Pull request merged**